src/chainstate/coordinator/mod.rs

```diff
@@ -500,20 +540,561 @@ fn dispatcher_announce_burn_ops<T: BlockEventDispatcher>(
+                                            sn.block_height,
+                                            &sn_pox_id
+                                        );
+                                        if affirmation_pox_id.has_prefix(&sn_pox_id) {
+                                            // have already processed this sortitoin
+                                            test_debug!("Already processed sortition {} at height {} with PoX ID {} on canonical affirmation map {}", &sn.sortition_id, sn.block_height, &sn_pox_id, &heaviest_am);
+                                            found = true;
+                                            last_invalidate_start_block = height;
+                                            valid_sortition_ids.push(sn.sortition_id);
+                                            break;
+                                        }
+                                    }
+                                    if !found && num_sns > 0 {
+                                        // there are snapshots, and they're all diverged
+                                        debug!("No snapshot at height {} has a PoX ID that is a prefix of {} (affirmation map {})", height, &affirmation_pox_id, &heaviest_am);
+                                        break;
+                                    }
+                                }
+                                break;
+                            }
+                        }
+
+                        if !found_diverged {
+                            continue;
+                        }
+
+                        // we may have processed some sortitions correctly within this reward
+                        // cycle. Advance forward until we find one that we haven't.
+                        info!(
+                            "Re-playing sortitions starting within reward cycle {} burn height {}",
+                            rc, last_invalidate_start_block
+                        );
+
+                        diverged = true;
+                        break;
+                    }
+
+                    if diverged {
+                        // find our ancestral sortition ID that's the end of the last reward cycle
+                        // the new affirmation map would have in common with the old affirmation
+                        // map, and invalidate its descendants
+                        let ic = self.sortition_db.index_conn();
+                        let sortition_id = self.canonical_sortition_tip.as_ref().expect(
+                            "FAIL: processing an affirmation reorg, but don't have a canonical sortition tip",
+                        );
+
+                        // first snapshot in which we'll invalidate all descendant snapshots, but retain some previously-invalidated snapshots
+                        let revalidate_sn = SortitionDB::get_ancestor_snapshot(
+                            &ic,
+                            first_invalidate_start_block - 1,
```

Comment:
It can't be zero, since its smallest value is the first burnchain block height.  It's `- 1` because we want to invalidate all sortitions that *descend* from `revalidate_sn`, which must be the *parent* of the first sortition we want to invalidate.